pkg/vault/test/foundry/E2eSwap.t.sol

```diff
@@ -0,0 +1,696 @@
+            );
+        } else {
+            // Accepts an error of 0.0000001% between amountIn from ExactOut and ExactIn swaps. This error is caused by
+            // differences in the computeInGivenExactOut and computeOutGivenExactIn functions of the pool math.
+            assertApproxEqRel(exactAmountInSwap, exactAmountIn, 1e9, "ExactOut and ExactIn amountsIn should match");
+        }
+    }
+
+    function testExactInRepeatExactOutVariableFees__Fuzz(uint256 exactAmountIn, uint256 poolSwapFeePercentage) public {
+        exactAmountIn = bound(exactAmountIn, minSwapAmountTokenA, maxSwapAmountTokenA);
+
+        poolSwapFeePercentage = bound(poolSwapFeePercentage, minPoolSwapFeePercentage, maxPoolSwapFeePercentage);
+        vault.manualSetStaticSwapFeePercentage(pool, poolSwapFeePercentage);
+
+        vm.startPrank(sender);
+        uint256 snapshotId = vm.snapshot();
+        uint256 exactAmountOut = router.swapSingleTokenExactIn(
+            pool,
+            tokenA,
+            tokenB,
+            exactAmountIn,
+            0,
+            MAX_UINT128,
+            false,
+            bytes("")
+        );
+
+        uint256 feesTokenB = vault.getAggregateSwapFeeAmount(pool, tokenB);
+
+        vm.revertTo(snapshotId);
+        uint256 exactAmountInSwap = router.swapSingleTokenExactOut(
+            pool,
+            tokenA,
+            tokenB,
+            exactAmountOut + feesTokenB,
+            MAX_UINT128,
+            MAX_UINT128,
+            false,
+            bytes("")
+        );
+        uint256 feesTokenA = vault.getAggregateSwapFeeAmount(pool, tokenA);
+        vm.stopPrank();
+
+        if (decimalsTokenA != decimalsTokenB) {
+            // If tokens have different decimals, an error is introduced in the computeBalance in the order of the
+            // difference of the decimals.
+            uint256 tolerance;
+            if (decimalsTokenA < decimalsTokenB) {
+                tolerance = 10 ** (decimalsTokenB - decimalsTokenA + 1);
+            } else {
+                tolerance = 10 ** (decimalsTokenA - decimalsTokenB + 1);
+            }
+            assertApproxEqAbs(
+                exactAmountInSwap - feesTokenA,
+                exactAmountIn,
+                tolerance,
+                "ExactOut and ExactIn amountsIn should match"
+            );
+        } else {
+            // Accepts an error of 0.0000001% between amountIn from ExactOut and ExactIn swaps. This error is caused by
+            // differences in the computeInGivenOut and computeOutGivenIn functions of the pool math.
+            assertApproxEqRel(
+                exactAmountInSwap - feesTokenA,
+                exactAmountIn,
+                1e9,
+                "ExactOut and ExactIn amountsIn should match"
+            );
+        }
+    }
+
+    function _checkUserBalancesAndPoolInvariant(
+        BaseVaultTest.Balances memory balancesBefore,
+        BaseVaultTest.Balances memory balancesAfter,
```

Comment:
One more to add could be live balances.
WDYT? Might be an overkill to test that here though. Not a strong opinion.

Reply:
actually, might make sense to run the existing test against tokens that are yield bearing. But that's another PR.

Reply:
I'll create a test with Yield-Bearing tokens in another PR.